src/components/lib/crt/crt.c

```diff
@@ -167,6 +217,74 @@ crt_comp_create_with(struct crt_comp *c, char *name, compid_t id, struct crt_com
 	cos_compinfo_init(cos_compinfo_get(c->comp_res),
 			  r->ptc, r->ctc, r->compc, r->heap_ptr, r->captbl_frontier,
 			  cos_compinfo_get(cos_defcompinfo_curr_get()));
+	return 0;
+}
+
+/** 
+ * Create the component from the checkpoint
+ * c is the new component
+ * chkpt is the checkpoint used to create c
+ * name is c's name
+ * id is c's component id
+ */
+int 
+crt_comp_create_from(struct crt_comp *c, char *name, compid_t id, struct crt_chkpt *chkpt)
+{
+	struct cos_compinfo *ci, *root_ci;
+	struct cos_component_information *comp_info;
+	unsigned long info_offset;
+	size_t  ro_sz,   rw_sz, data_sz, bss_sz;
+	char   *ro_src, *data_src, *mem;
+	int     ret;
+	vaddr_t	info = chkpt->c->info;
+
+	assert(c && name);
+
+	if (crt_comp_init(c, name, id, NULL, chkpt->c->info)) BUG();
+	ci      = cos_compinfo_get(c->comp_res);
+	root_ci = cos_compinfo_get(cos_defcompinfo_curr_get());
+
+	// overwrite
+	c->entry_addr = chkpt->c->entry_addr;
+	c->ro_addr = chkpt->c->ro_addr;
+	c->rw_addr = chkpt->c->rw_addr;
+
+	//re-work the sinvs with the right IDs
+	memcpy(c->sinvs, chkpt->c->sinvs, sizeof(c->sinvs));
+	c->n_sinvs = chkpt->c->n_sinvs;
+	for(u32_t i = 0; i < c->n_sinvs; i++) {
+		//if the client was the old component, update id in new
+		if(chkpt->c->sinvs[i]->client->id == chkpt->c->id) {
+			c->sinvs[i]->client = c;
+		}
+		if(chkpt->c->sinvs[i]->server->id == chkpt->c->id) {
+			c->sinvs[i]->server = c;
+		}
+	}
```

Comment:
Isn't all of this creating pointers to the same set of component sinvs? Since the sinvs array is an array of pointers, aren't all checkpoints of the same component going to point to the same actual `crt_sinv`s? Thus any updates we make to the ids will be overwritten by each checkpoint?

Reply:
If this is true, the alternative is to just have the sinvs array in each chkpoint/component be of actual `crt_sinv`s rather than pointers.